### ISSUE TITLE
Inference not working for union without discriminant properties

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#34590]
**TypeScript Version:**  3.6.3

**Search Terms:**

**Code**

```ts
type AElement = HTMLAnchorElement;
type BElement = HTMLButtonElement;

interface A {
    a: 'a';
    onClick(el: AElement): void;
}

interface B {
    onClick(el: BElement): void;
}

type Props = A | B;
const props: Props = {
    a: 'a',
    // unexpected error, `el` is implicitly `any`
    onClick: el => {}
};
```

If we add a common discriminant property, it does work:

```ts
type AElement = HTMLAnchorElement;
type BElement = HTMLButtonElement;

interface A {
    type: 'a';
    onClick(el: AElement): void;
}

interface B {
    type: 'b';
    onClick(el: BElement): void;
}

type Props = A | B;
const props: Props = {
    type: 'a',
    onClick: el => {},
};
```

It will also work if the union members do not share the same property:

```ts
type AElement = HTMLAnchorElement;

interface A {
    a: 'a';
    onClick(el: AElement): void;
}

interface B {
    foo: number;
}

type Props = A | B;
const props: Props = {
    a: 'a',
    onClick: el => {},
};
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:** 

**Related Issues:**
- https://github.com/microsoft/TypeScript/issues/20863
- https://github.com/Microsoft/TypeScript/issues/30631
- https://github.com/microsoft/TypeScript/issues/29025#issuecomment-543922052

**Comment by j-oliveras** [comment]
As a workaround of your first code (add a optional property `a` with type undefined):
```ts
type AElement = HTMLAnchorElement;
type BElement = HTMLButtonElement;

interface A {
    a: 'a';
    onClick(el: AElement): void;
}

interface B {
    a?: undefined;
    onClick(el: BElement): void;
}

type Props = A | B;
const props: Props = {
    a: 'a',
    onClick: el => {}
};
```

[Playgorund](https://www.typescriptlang.org/play/index.html?ts=3.7-Beta&ssl=1&ssc=1&pln=18&pc=3#code/C4TwDgpgBAggogGwgWwgO2FAvFAEgFQFkAZGNAYwAsB7AJ0RXWAG4BYAKFEigCEHUM2PEWI8ArsGDU0-Jm3YcAlhgi0AZgENy0GFADeHKEagaAXFADkGi-ONRpAYQSLyAawAUEBOfhIBwAEpzADdqRQATeQBfDiUVdS1oHn1DYw0AfnMxNHCINWUISNSjR2c3T29eWQwgqFCI6NjOcGgABVpqMABnIV0AH155cmkuzDAO7vN2zp6cA3Y7M0trABpi+zQnF1dzL2wAPn0Y9ijmIA) (tested with 3.7-beta, 3.6.3 and nightly).

**Comment by RyanCavanaugh** [comment]
I don't understand the intended use of this code. When some code that receives a `Props` goes to call `onClick`, how does it determine what to do? What about aliasing? How do you get into this state in a way where callers are known to do the right thing? It seems like a correct error to say that TS can't really predict whether or not `onClick` will be invoked with an `AElement` or `BElement` since there's no runtime way to safely distinguish these objects.

**Comment by OliverJAsh** [comment]
> When some code that receives a Props goes to call onClick, how does it determine what to do?
Ideally I could do this:

```ts
declare const aElement: AElement;
declare const bElement: BElement;
props.a === 'a' ? props.onClick(aElement) : props.onClick(bElement);
```

Although as I just discovered, this won't work.

> since there's no runtime way to safely distinguish these objects.

Is that because `a: 'a'` _could_ exist on `B`? In that case, I guess what I'm really after is exact types?

**Comment by ferdaber** [comment]
In the context of usage, this is used for polymorphic components in React (where the most popular use case is rendering buttons vs anchors that visually look the same).

If there is a `href` prop given to this polymorphic component, then the component will render an anchor tag, otherwise it will render a button tag, so the discriminator is the presence of a property in the given props.

**Comment by karol-majewski** [comment]
We could use solid discriminants or `a?: undefined`/`a?: never` for `B` as @j-oliveras suggested, so at least we have a way out @OliverJAsh 👍

**Comment by RyanCavanaugh** [comment]
The problem with a nonpresent discriminant is that it's subject to aliasing effects:

```ts
type AElement = HTMLAnchorElement;
type BElement = HTMLButtonElement;

interface A {
    a: 'a';
    onClick(el: AElement): void;
}

interface B {
    onClick(el: BElement): void;
}

type Props = A | B;

function fail(x: A | B) {
  if (x.a === "a") {
    x.onClick(someAnchor);
  }
}

// Clearly OK, by definition
let tricky1 = { a: "a", onClick: (el: BElement) => e.buttonThing() };
// Clearly OK - 'a' property is aliased away, 'onClick' property matches
let tricky2: B = tricky1;
// Clearly OK by the definition of union types
let tricky3: A | B = tricky2;
// Crashes
fail(tricky3);
```
This is why you need an `undefined` or `never` placeholder property to make the assignment into `tricky2` illegal

**Comment by ferdaber** [comment]
For use cases of polymorphic components, overload signatures will work fine but they will break in cases where a conditional type attempts to infer its props, so I was hoping this would be a solution but your reasoning makes sense to me.

**Comment by SlurpTheo** [comment]
> ```ts
> // Clearly OK by the definition of union types
> let tricky3: A | B = tricky2;
> ```

Blocking this (w/o having to manually sprinkle `never` props) is [another](https://github.com/microsoft/TypeScript/issues/20863#issuecomment-539676594) place that could benefit from [ts-toolbelt](https://github.com/pirix-gh/ts-toolbelt)'s `Union.Strict<T>`:
```ts
import { Union } from "ts-toolbelt";
const tricky4 = { a: "a", onClick: (el: BElement) => el.checkValidity() };
const tricky5: Union.Strict<A | B> = tricky4; // ts(2322)
```

![image](https://user-images.githubusercontent.com/5269716/67580623-3c9b7500-f70c-11e9-94a3-ac00324b7487.png)
